drivers/adc/adc_mchp_xec.c

```diff
@@ -0,0 +1,310 @@
+	u32_t delay_reg;
+	u32_t status_reg;
+	u32_t single_reg;
+	u32_t repeat_reg;
+	u32_t channel_read_reg[8];
+	u32_t unused[18];
+	u32_t config_reg;
+	u32_t vref_channel_reg;
+	u32_t vref_control_reg;
+	u32_t sar_control_reg;
+};
+
+#define ADC_XEC_REG_BASE						\
+	((struct adc_xec_regs *)(DT_INST_0_MICROCHIP_XEC_ADC_BASE_ADDRESS))
+
+
+DEVICE_DECLARE(adc_xec);
+
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct adc_xec_data *data = CONTAINER_OF(ctx, struct adc_xec_data, ctx);
+	struct adc_xec_regs *adc_regs = ADC_XEC_REG_BASE;
+
+	data->repeat_buffer = data->buffer;
+
+	adc_regs->single_reg = ctx->sequence.channels;
+	adc_regs->control_reg |= XEC_ADC_CTRL_START_SINGLE;
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx,
+					      bool repeat_sampling)
+{
+	struct adc_xec_data *data = CONTAINER_OF(ctx, struct adc_xec_data, ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->repeat_buffer;
+	}
+}
+
+static int adc_xec_channel_setup(struct device *dev,
+				 const struct adc_channel_cfg *channel_cfg)
+{
+	struct adc_xec_regs *adc_regs = ADC_XEC_REG_BASE;
+	u32_t reg;
+
+	ARG_UNUSED(dev);
+
+	if (channel_cfg->acquisition_time != ADC_ACQ_TIME_DEFAULT) {
+		return -EINVAL;
+	}
+
+	if (channel_cfg->channel_id >= MCHP_ADC_MAX_CHAN) {
+		return -EINVAL;
+	}
+
+	if (channel_cfg->gain != ADC_GAIN_1) {
+		return -EINVAL;
+	}
+
+	/* Setup VREF */
+	reg = adc_regs->vref_channel_reg;
+	reg &= ~MCHP_ADC_CH_VREF_SEL_MASK(channel_cfg->channel_id);
+
+	if (channel_cfg->reference == ADC_REF_INTERNAL) {
+		reg |= MCHP_ADC_CH_VREF_SEL_PAD(channel_cfg->channel_id);
+	} else if (channel_cfg->reference == ADC_REF_EXTERNAL0) {
+		reg |= MCHP_ADC_CH_VREF_SEL_GPIO(channel_cfg->channel_id);
+	} else {
+		return -EINVAL;
+	}
+
+	adc_regs->vref_channel_reg = reg;
+
+	/* Differential mode? */
+	reg = adc_regs->sar_control_reg;
+	reg &= ~BIT(MCHP_ADC_SAR_CTRL_SELDIFF_POS);
+	if (channel_cfg->differential != 0) {
+		reg |= MCHP_ADC_SAR_CTRL_SELDIFF_EN;
+	}
+	adc_regs->sar_control_reg = reg;
+
+	return 0;
+}
+
+static bool adc_xec_validate_buffer_size(const struct adc_sequence *sequence)
+{
+	int chan_count = 0;
+	size_t buff_need;
+	u32_t chan_mask;
+
+	for (chan_mask = 0x80; chan_mask != 0; chan_mask >>= 1) {
+		if (chan_mask & sequence->channels) {
+			chan_count++;
+		}
+	}
+
+	buff_need = chan_count * sizeof(u16_t);
+
+	if (sequence->options) {
+		buff_need *= 1 + sequence->options->extra_samplings;
+	}
+
+	if (buff_need > sequence->buffer_size) {
+		return false;
+	}
+
+	return true;
+}
+
+static int adc_xec_start_read(struct device *dev,
+			      const struct adc_sequence *sequence)
+{
+	struct adc_xec_regs *adc_regs = ADC_XEC_REG_BASE;
+	struct adc_xec_data *data = dev->driver_data;
+	u32_t reg;
+
+	if (sequence->channels & ~MCHP_ADC_MAX_CHAN_MASK) {
+		return -EINVAL;
+	}
+
+	if (sequence->channels == 0UL) {
+		return -EINVAL;
+	}
+
+	if (!adc_xec_validate_buffer_size(sequence)) {
+		return -ENOMEM;
+	}
+
```

Comment:
@dcpleung any reason to not implement adc continuous read and handling adc_sequence_options ?

Reply:
Do you mean repeat read in hardware? The API expects reading of all channels one time per callback, and it sets up another read again. So doing repeat reading in hardware would cause errors in terms of ADC context. As for the sequence options, they are handled by the ADC context (mostly). We just need to make sure the values are sane.